### PR TITLE
core/mutex: Add mutex_cancel

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This commit adds the (pseudo)-module `mutex_cancel`, which unlocks the likewise named function. With this function, a thread blocked with `mutex_lock()` can be unblocked without it obtaining the mutex. A caller of `mutex_lock()` is notified via an `-ECANCELED` return code.

This is intended to simplify the implementation of `xtimer_mutex_lock_timeout()` and to implement `ztimer_mutex_lock_timeout()`.

### Testing procedure

With `tests/mutex_cancel` a test application is provided that must succeed.

### Issues/PRs references

Split out of https://github.com/RIOT-OS/RIOT/pull/15052

### CODE REVIEW

File: core/mutex.c [hunk 1]
```diff
@@ -131,3 +147,31 @@ void mutex_unlock_and_sleep(mutex_t *mutex)
...
+    unsigned irq_state = irq_disable();
+    if (thread->status >= STATUS_ON_RUNQUEUE) {
+        thread->wait_data = mutex;
+        irq_restore(irq_state);
+        return;
+    }
+
+    if ((mutex->queue.next != MUTEX_LOCKED) &&
+            (mutex->queue.next != NULL) &&
+            list_remove(&mutex->queue, (list_node_t *)&thread->rq_entry)) {
```
Comment (benpicco) [inline core/mutex.c:163]:
```suggestion
    if ((mutex->queue.next != MUTEX_LOCKED)
        && (mutex->queue.next != NULL)
        && list_remove(&mutex->queue, (list_node_t *)&thread->rq_entry)) {
```

looks a bit more tidy, but that's just my personal taste

**Review by benpicco — COMMENTED** [review]
That's pretty neat!
Just the other day I was wondering how to deliver a message to a thread that is waiting for an interrupt with `mutex_lock(&m)` - now I can just 

```C
if (mutex_lock(&m) < 0) {
    msg_receive(&msg);
    …
}
```

Travis is not happy btw.

File: core/mutex.c [hunk 2]
```diff
@@ -49,6 +49,13 @@ int mutex_lock(mutex_t *mutex)
     }
 
     thread_t *me = thread_get_active();
+#if IS_USED(MODULE_CORE_MUTEX_CANCEL)
+    if (me->wait_data == mutex) {
```
[the remark this answers is not in the document]
Comment (maribu) [reply core/mutex.c:53]: Yes, indeed. I will add doc next week

File: core/mutex.c [hunk 3]
```diff
@@ -60,10 +67,19 @@ int mutex_lock(mutex_t *mutex)
         thread_add_to_list(&mutex->queue, me);
     }
 
+#if IS_USED(MODULE_CORE_MUTEX_CANCEL)
+    me->wait_data = NULL;
```
[the remark this answers is not in the document]
Comment (maribu) [reply core/mutex.c:71]: Exactly.

File: core/mutex.c [hunk 4]
```diff
@@ -60,10 +67,19 @@ int mutex_lock(mutex_t *mutex)
...
+#endif
 
     irq_restore(irq_state);
     thread_yield_higher();
     /* We were woken up by scheduler. Waker removed us from queue. */
+#if IS_USED(MODULE_CORE_MUTEX_CANCEL)
+    if (me->wait_data) {
```
[the remark this answers is not in the document]
Comment (maribu) [reply core/mutex.c:78]: Yes. At this point we can be sure that wait data is only set (to the address of the mutex) as a signal to the cancellation. Before, a non-NULL value could also be a leftover from another user of the `wait_data`. At least I believe it os safe to do so hear (and often one instruction less to compare against zero). I'll add doc accordingly.

**Comment by maribu** [comment]
I have to double check that using `wait_data` is safe in conjunction with `thread_flags`. If `wait_data` could end up pointing to the mutex by chance, it would be pretty catastrophic.

**Comment by kaspar030** [comment]
I would like some comment on when to use this. Maybe we should keep this somewhat internal, as in, advertise as supposed to be used only by system libraries or similar?

**Comment by maribu** [comment]
> Maybe we should keep this somewhat internal, as in, advertise as supposed to be used only by system libraries or similar?

I don't think this will ease maintenance, if that is the intention. IMO it doesn't matter if we have to maintain this features due to it being used `xtimer_mutex_lock_timeout()` and `ztimer_mutex_lock_timeout()`, or because it is directly supported as public API.

In fact, the semantics of this function originally where different, but I had to change it in order to allow `xtimer_mutex_lock_timeout()` to be implemented on it. This is IMO a good indicator that the semantics is already set in stone due to the feature being exposed indirectly. So why not making it officially set in stone then.

(Originally only a currently blocked threads could be unblocked without obtaining the mutex, no "cancel ahead of time" was allowed. This was significantly easier to implement and faster. However, `xtimer_mutex_lock_timeout()` explicitly allows a timeout below `XTIMER_BACKOFF` and this is even checked in the unit test. IMO it would have been best to just fall back to `mutex_trylock()` for timeouts below `XTIMER_BACKOFF` and don't even bother with calling into `xtimer` at all for zero or near zero timeouts. But the semantics of `xtimer_mutex_lock_timeout()` have already been settled and I adapted `mutex_cancel()` to allow implementing this.)

**Comment by kaspar030** [comment]
> In fact, the semantics of this function originally where different, but I had to change it in order to allow `xtimer_mutex_lock_timeout()` to be implemented on it. This is IMO a good indicator that the semantics is already set in stone due to the feature being exposed indirectly. So why not making it officially set in stone then.

Having the option to cancel a mutex makes implementation of xtime_mutex_lock_timeout() et al easier, I like that. But a basic primitive like mutex _usually_ doesn't have a cancel method, and _usually_ doesn't need its return code checked for "lock() exited, is the mutex now locked?".

In C, ensuring that code won't break because suddenly mutexes are canceled is usually done by convention ("don't call mutex_cancel() unless all users of that mutex expect that!". That's fine, C world works like this. But these things also make reasoning about code more difficult, because there are now constraints that are not in code. In our case, mutex is one of the most used primitives, and exposing mutex_cancel() makes all that code essentially prone to undefined behavior. It is also not feasible to check the return code of mutex_unlock() everywhere.

There has been some work on providing safe Rust abstractions on top of RIOT's API, e.g., @chrysn's awesome [riot-wrappers](https://gitlab.com/etonomy/riot-wrappers). I'm also working on a drop in replacement for core written in Rust.
In Rust, it is possible to express many more constraints in actual code, so the compiler can prevent a lot of misuse.
Now, when mutex_lock() can suddenly return "ECANCELED", the guarantees of those higher-level abstractions become watered, as in, they depend on other code using the mutex to not call mutex_cancel(), (even if that is unlikely to happen). If the concurrency guarantees are to be held at compile time, that case needs to be handled, and it needs to be handled *where it might show up* (so not once, but in code locking mutexes) . Going with convention here would make the abstractions less safe, 
handling the case in order to keep guarantees would make the abstractions less convenient and less efficient.

Previously, we had core and xtimer intermingled badly at the code level (xtimer accessed core structs directly and had half a mutex implementation inside). That was obviously suboptimal.
With this PR, the mutex implementation becomes intermingled with xtimer at API level. I prefer that over the previous solution, but it requires mutex_lock() to possibly return ECANCELED.

As long as the use of this is contained to mutex+xtimer+ztimer, higher level code can pretty safely just not handle ECANCELED outside of the mutex lock timeout functions. If mutex_cancel() is part of the public API, higher level code needs to handle the case, or be open to misuse.

If it were to me, I'd put the function declaration in a separate header that is not supposed to be included by "normal" mutex users.
Or just re-declare in xtimer.c and ztimer_core.c.
But I'd be fine with doing this through documentation.

> IMO it doesn't matter if we have to maintain this features due to it being used xtimer_mutex_lock_timeout() and ztimer_mutex_lock_timeout(), or because it is directly supported as public API.

Asked the other way around - do we know any other use case for mutex_cancel()?
If we don't, how about we keep it out of the public API for now, and re-visit once we do?
That would make life easier for people writing safe higher-level abstractions.

**Comment by chrysn** [comment]
I've just only read through this being pinged -- what I don't quite understand from a first look at this is: Why does this need something changed around mutexes? Given that we have the mechanism of waiting for thread flags, can't a function like "lock this mutex or time out" just be implemented using `thread_flags_wait_one(THREAD_FLAG_MUTEX_READY | THREAD_FLAG_TIMEOUT)`?

(And I too see trouble in making a function that previously worked unconditionally suddenly fallible, but I may still lack parts of the big picture).

**Comment by benpicco** [comment]
This certainly allows for some neat hacks, but also dirty ones.
I thought it would be neat for being able to inject shell commands into the shell thread.
Right now it will wait on a mutex in `isrpipe_read()`.

With this we could do (roughly) 

```patch
--- a/sys/shell/shell.c
+++ b/sys/shell/shell.c
@@ -470,6 +470,12 @@ void shell_run_once(const shell_command_t *shell_commands,
     while (1) {
         int res = readline(line_buf, len);
 
+        msg_t m;
+        if (msg_try_receive(&m) == 1) {
+            handle_input_line(shell_commands, m.content.ptr);
+            continue;
+        }
+
         switch (res) {
 
             case EOF:
```

Granted, that *feels* dirty - and I know why: A thread might wait on multiple mutexen. How can we be sure to cancel the right on?

**Comment by kaspar030** [comment]
> Granted, that _feels_ dirty - and I know why: A thread might wait on multiple mutexen. How can we be sure to cancel the right on?
> Maybe the caller as to supply the pointer to the mutex it wants to cancel?

That's how @maribu did it in this PR:

https://github.com/RIOT-OS/RIOT/pull/15442/files#diff-bb070320207448eaf735fc594635be8523905aac140c757d00d43581d6e2cfcdR225-R242

I agree that your example feels somewhat dirty. :)

Maybe we should change stdio/uart/isr_pipe to use thread flags, so waiting for them can be done at the same time as waiting for other flags, timeouts, msg, events, ... which would allow this to be implemented in a clean way.

**Comment by kaspar030** [comment]
Anyhow, let's not block this unnecessarily. Is it so bad to restrict the function for now by adding some comment?

**Comment by maribu** [comment]
@chrysn:

> Given that we have the mechanism of waiting for thread flags, can't a function like "lock this mutex or time out" just be implemented using `thread_flags_wait_one(THREAD_FLAG_MUTEX_READY | THREAD_FLAG_TIMEOUT)`?

That would be better, but it will sadly not work in general. E.g. let's say we have

``` C
extern mutex_t mut_a, mut_b;

void * thread_impl(void *arg) {
    while (1) {
        if (mutex_lock(&mut_a) < 0) {
            continue;
        }
        if (mutex_lock(&mut_b) == 0) {
            perform_task(a);
            mutex_unlock(&mut_b);
        }
        mutex_unlock(&mut_a);
    }
}
```

In that case, the thread could either be blocked waiting for `mut_a` or for `mut_b`. We need to make sure that we don't cancel the wrong mutex here. We cannot assume that the threads are somehow synchronized - after all the mutex is provided as a tool to synchronize threads.

(Agreed, your argument was in the context of the `xtimer_mutex_lock_timeout()` use case. But this problem applies there just as well.)

**Comment by maribu** [comment]
@kaspar030: How about making `mutex_lock()` return `void` and provide `mutex_lock_may_fail()` in addition? Code not checking the return code of `mutex_lock_may_fail()` would be an obvious red flag.

This would likely yield a bit more ROM use, but this would also get rid of the performance penalty to `mutex_lock()` that using the `core_mutex_cancel` module results in. To me, it makes much more sense if that performance penalty is only paid by callers of `mutex_lock_may_fail()`.

**Comment by kaspar030** [comment]
> How about making `mutex_lock()` return `void` and provide `mutex_lock_may_fail()` in addition? Code not checking the return code of `mutex_lock_may_fail()` would be an obvious red flag.

That sounds nice. We can even mark that as "result must be checked" so the compiler warns!

**Comment by kaspar030** [comment]
but, what happens when the lock gets canceled anyways? (locker uses "mutex_lock()", but someone calls "mutex_cancel()" on it?

**Comment by maribu** [comment]
> but, what happens when the lock gets canceled anyways? (locker uses "mutex_lock()", but someone calls "mutex_cancel()" on it?

`mutex_lock()` must never fail :-)

Btw: I think I will place a preconditions that `mutex_lock_may_fail()` is called with IRQs disabled. This would get rid of all the complexity, as one could simply do:

```C
   setup_isr_that_unblocks_mutex_but_mask_it();
   unsigned state = irq_disable();
   unmask_isr_that_unblocks_mutex();
   if (mutex_lock_may_fail(&mut, state) == 0) {
       work_on_shared_data();
       mutex_unlock();
   }
```

Here there is no chance that the ISR hits right before the call to `mutex_lock_may_fail()`, and all those abuse `wait_data` to signal cancellation is no longer needed. So not even a pseudo-module would be required to enable it, as the linker would be totally capable to garbage collect every single byte of RAM and ROM needed by this.

**Comment by benpicco** [comment]
> Btw: I think I will place a preconditions that `mutex_lock_may_fail()` is called with IRQs disabled. This would ~~get rid of all the complexity~~ move the complexity to the caller

The mutex might as well be canceled by another thread.

**Comment by maribu** [comment]
> The mutex might as well be canceled by another thread.

Context switches are either due to the running thread yielding / blocking, or due to an IRQ. So it won't happen with IRQs disabled by surprise.

**Comment by benpicco** [comment]
Yes but I meant that

```C
   setup_isr_that_unblocks_mutex_but_mask_it();
   unsigned state = irq_disable();
   unmask_isr_that_unblocks_mutex();
```

procedure.

If you have no interrupt to configure, there would be no difference to the current `mutex_lock()` which disables interrupts as well.

**Comment by kaspar030** [comment]
> Btw: I think I will place a preconditions that `mutex_lock_may_fail()` is called with IRQs disabled.

how do you unblock specific IRQs in RIOT?

IMO this is getting out of hand. Why not just keep mutex_unlock() private to mutex, xtimer, ztimer?

File: core/mutex.c [hunk 5]
```diff
@@ -131,3 +147,31 @@ void mutex_unlock_and_sleep(mutex_t *mutex)
     irq_restore(irqstate);
     thread_yield_higher();
 }
+
+#if IS_USED(MODULE_CORE_MUTEX_CANCEL)
+void mutex_cancel(mutex_t *mutex, thread_t *thread)
+{
+    unsigned irq_state = irq_disable();
+    if (thread->status >= STATUS_ON_RUNQUEUE) {
```
[the remark this answers is not in the document]
Comment (maribu) [reply core/mutex.c:189]: Done (slightly different)

**Comment by maribu** [comment]
The last fixup is to make C++ happy. Apparently, readable C code is something C++ doesn't like :-/

**Comment by fjmolinas** [comment]
@kaspar030 seems like @maribu has made some changes since the last time you looked!

File: core/mutex.c [hunk 6]
```diff
@@ -33,21 +33,20 @@
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
-int mutex_lock(mutex_t *mutex)
+/**
+ * @brief   Block waiting for a locked mutex
+ * @pre     IRQs are disabled
```
Comment (kaspar030) [inline core/mutex.c:38]: if this is a pre, why is the irq_state necessary? just hard enable IRQs afterwards.

**Comment by maribu** [comment]
> @kaspar030 seems like @maribu has made some changes since the last time you looked!

Yes. Almost a rewrite. Should contain only the usual amount footguns present in all code written in C.

File: core/mutex.c (hunk 6 above)
Comment (maribu) [reply core/mutex.c:38]: It's not :-/ Sneaked in.

Comment (maribu) [reply core/mutex.c:38]: Wait, this is right. I just saw the `mutex_lock` and misread as if I added this to `mutex_lock()` somehow.

And `irq_state` is need in case someone does stuff like:

```C
    unsigned state = irq_disable();
    mutex_lock(&mut);
    other_stuff_with_irqs_disabled_i_know_this_is_stupid();
    irq_restore(state);    
```

I don't this that there is a valid use case for the above code. But IMO we shouldn't judge what application developers implement. And if the want to do stupid stuff, the should certainly be able to do stupid stuff in a safe manner.

File: core/mutex.c [hunk 7]
```diff
@@ -33,21 +33,20 @@
...
+ * @post    IRQs are enabled
```
Comment (benpicco) [inline core/mutex.c:39]: so this is not true

Comment (maribu) [reply core/mutex.c]: Yes, fixed. Thanks!

File: core/mutex.c (hunk 6 above)
Comment (kaspar030) [reply core/mutex.c:38]: > I don't think that there is a valid use case for the above code.

no, it will probably horribly fail, as the mutex code expects to schedule away on irq_restore(). But that is not new to this PR: :/

File: core/mutex.c [hunk 8]
```diff
@@ -60,11 +59,59 @@ int mutex_lock(mutex_t *mutex)
...
     thread_yield_higher();
     /* We were woken up by scheduler. Waker removed us from queue. */
-    return 0;
+}
+
+void mutex_lock(mutex_t *mutex)
+{
+    unsigned irq_state = irq_disable();
+
+    DEBUG("PID[%" PRIkernel_pid "]: Mutex in use.\n", thread_getpid());
```
Comment (kaspar030) [inline core/mutex.c:71]: What does this mean? Maybe rephrase: "pid x: Locking mutex %p"?

**Review by kaspar030 — COMMENTED** [review]
As far as I can see, regular mutex usage is unaffected.
I tested master/PR, code size is identical. "tests/bench_mutex_pingpong" shows 2 ticks more for the PR (on Cortex-M4), which can be considered identical.

I did not yet look into detail for the mutex_cancel() implementation.
(@maribu if you want to split PR, I'd ACK the factoring-out of _block()).

File: core/include/mutex.h [hunk 9]
```diff
@@ -126,6 +126,18 @@ typedef struct {
     list_node_t queue;
 } mutex_t;
 
+/**
+ * @brief   A cancellation structure for with with @ref mutex_lock_cancelable
```
Comment (miri64) [inline core/include/mutex.h:130]:
```suggestion
 * @brief   A cancellation structure for @ref mutex_lock_cancelable
```
?

File: core/mutex.c (hunk 6 above)
Comment (maribu) [reply core/mutex.c:38]: Fixing that would finally be a use case for the return value of `irq_enable()` :-D

File: core/include/mutex.h (hunk 9 above)
Comment (maribu) [reply core/include/mutex.h:130]: Good catch. I think I meant "for use with".

File: core/mutex.c (hunk 8 above)
Comment (maribu) [reply core/mutex.c]: I made the debug output more consistent and just use the function names, which both describe quite well what is done and how to find the corresponding code.

File: core/mutex.c [hunk 10]
```diff
@@ -60,19 +59,68 @@ int mutex_lock(mutex_t *mutex)
...
+}
+
+int mutex_lock_cancelable(mutex_cancel_t *mc)
+{
+    unsigned irq_state = irq_disable();
+
+    DEBUG("PID[%" PRIkernel_pid "]: mutex_lock_cancelable()\n",
+          thread_getpid());
+
+    if (mc->cancelled) {
```
Comment (benpicco) [inline core/mutex.c:92]:
```suggestion
    if (mc->cancelled) {
        mc->cancelled = 0;
```

so we can re-use the mutex and cancel it again

File: core/mutex.c [hunk 11]
```diff
@@ -60,19 +59,68 @@ int mutex_lock(mutex_t *mutex)
...
+        /* mutex is unlocked. */
+        mutex->queue.next = MUTEX_LOCKED;
+        DEBUG("PID[%" PRIkernel_pid "]: mutex_lock_cancelable() early out.\n",
+              thread_getpid());
+        irq_restore(irq_state);
+        return 0;
+    }
+    else {
+        _block(mutex, irq_state);
+        if (mc->cancelled) {
```
Comment (benpicco) [inline core/mutex.c:110]: same here

**Review by benpicco — COMMENTED** [review]
looks good to me.
Please squash, having this separated into logical commits will make the changes clearer - I think we might merge this without splitting, as the changes to `mutex_lock()` are just moving code into a helper function.

File: core/mutex.c (hunk 10 above)
Comment (maribu) [reply core/mutex.c:92]: This won't work. The usual use case looks like this:

```C
static void timeout_cb(void *_arg) {
    mutex_cancel(arg);
}

int ztimer_mutex_lock_timeout(ztimer_clock_t *clock, mutex_t *mutex,
                              uint32_t timeout)
{
    mutex_cancel_t mc = mutex_cancel_init(mutex);
    ztimer_t t;
    t.callback = timeout_cb;
    t.arg = &mc;
    ztimer_set(clock, &t, timeout);
    if (0 == mutex_lock_cancelable(mutex)) {
        /* BANG! */
        ztimer_remove(clock, &t);
        return 0;
    }
    return -ECANCELED;
}
```

Let's say the timer expired at `/* BANG! */`. In that case `mc->cancel` will be set to `1` and `mutex_lock_cancelable()` has no chance to reset it to zero. The `mutex_cancel_t` structure must always be reset only after the IRQ triggering the cancel is disabled (or the thread sending the cancel is notified to no longer cancel the mutex).

Comment (benpicco) [reply core/mutex.c:92]: Ok that makes sense. So `mutex_cancel_init()` marks the earliest time a mutex can be canceled (also for subsequent calls)

**Review by benpicco — APPROVED** [review]
Looks good to me.
This makes for a useful addition. Having a way to abort a thread blocking on a mutex is a good tool to have.

I was contemplating a micro-optimization of just storing `-ECANCELED` in the `cancelled` flag and then unconditionally returning that, but it's cleaner this way.

Murdock is not happy btw.

File: core/mutex.c [hunk 12]
```diff
@@ -148,3 +181,33 @@ void mutex_unlock_and_sleep(mutex_t *mutex)
...
 }
+
+void mutex_cancel(mutex_cancel_t *mc)
+{
+    unsigned irq_state = irq_disable();
+    mc->cancelled = 1;
+
+    mutex_t *mutex = mc->mutex;
+    thread_t *thread = mc->thread;
+    if (thread->status >= STATUS_ON_RUNQUEUE) {
```
Comment (benpicco) [inline core/mutex.c:192]:
```suggestion
    if (thread_is_active(thread)) {
```

Comment (maribu) [reply core/mutex.c]: Done and also squashed right in

**Comment by fjmolinas** [comment]
@kaspar030  since you have taken a look at this one, can we get a second ACK?

**Comment by fjmolinas** [comment]
I've re-read through this PR discussion and from:

https://github.com/RIOT-OS/RIOT/pull/15442#issuecomment-727555354
> If it were to me, I'd put the function declaration in a separate header that is not supposed to be included by "normal" mutex users.
> Or just re-declare in xtimer.c and ztimer_core.c.
> But I'd be fine with doing this through documentation.

How different is what is being implemented here from a semaphore? Because adding a `mutex_cancel` module that is basically a semaphore just to make implementing `%timer_mutex_timeout` easier seems like a weird approach, and more of an argument to keep those private.

If adding the module is really justified, Hhow about having a separate module then? Like what is done for `event_%`? IMO that would be the best way of isolating what the vanilla mutex  from this one.

**Comment by maribu** [comment]
> How different is what is being implemented here from a semaphore?

Well, a mutex always fully equivalent to a semaphore with an initial value of 1 - regardless of this PR. However, this special case is so common that a separate implementation exploiting the optimization opportunities of this special case and a dedicated API is commonly considered to be sensible.

But anyway, IMO this discussion is off topic for this PR.

> how about having a separate module then?

We previously have identified splitting mutex code between `core` and `sys` as a bad practice, as this makes the code much less readable. E.g. when refactoring `core/mutex`, one would have to adapt `sys/mutex_cancel` in lockstep to avoid breakage. Those two features inherently rely on implementation details of each other, e.g. the format how a mutex is represented in memory. If one would refactor something in `core/mutex`, one would easily miss adapting `sys/mutex_cancel` and end up with subtle run time bugs.

Also: `mutex_lock()` and `mutex_lock_cancelable()` currently share code. This wouldn't be possible when trying to operate `mutex_lock_cancelable()` out of `core/mutex`.

**Comment by fjmolinas** [comment]
> Also: mutex_lock() and mutex_lock_cancelable() currently share code. This wouldn't be possible when trying to operate mutex_lock_cancelable() out of core/mutex.

Ok, good enough argument for me.

> We previously have identified splitting mutex code between core and sys as a bad practice, as this makes the code much less readable. E.g. when refactoring core/mutex, one would have to adapt sys/mutex_cancel in lockstep to avoid breakage. Those two features inherently rely on implementation details of each other, e.g. the format how a mutex is represented in memory. If one would refactor something in core/mutex, one would easily miss adapting sys/mutex_cancel and end up with subtle run time bugs.

I did not mean that the module would not be in core.

> However, this special case is so common that a separate implementation exploiting the optimization opportunities of this special case and a dedicated API is commonly considered to be sensible.

What are the other use cases besides `xtimer_mutex_lock_timeout()` and  `ztimer_mutex_lock_timeout()`? Because from what I understand one of the two is bound to be deprecated at some point which leaves only one real user.

Anyway I want to unblock this, I have some comments on the documentation but it should be good to go afterwards see below.

**Review by fjmolinas — COMMENTED** [review]
Some documentation remarks.

**Comment by benpicco** [comment]
> What are the other use cases besides `xtimer_mutex_lock_timeout()` and `ztimer_mutex_lock_timeout()`?

I think that function can be useful in general.
e.g. a function is blocking on a mutex to wait for an interrupt, but you want to abort the action.

My use case was waiting for new data on UART by blocking on a mutex, but being able to inject commands from outside and unblocking the mutex for that. 

You could argue that messages would also work here, but then you have the issue of overflowing message queues when all you want to signal is a single event.

File: core/include/mutex.h [hunk 13]
```diff
@@ -215,6 +261,68 @@ void mutex_unlock(mutex_t *mutex);
...
+ * @param[in,out]   mc      Mutex cancellation structure referring to the
+ *                          thread calling @ref mutex_lock_cancelable and to
+ *                          the mutex to cancel the operation on
+ *
+ * @pre     @p mc is use to cancel at most one call to
+ *          @ref mutex_lock_cancelable. (You can reinitialize the same memory
+ *          to safely reuse it.)
+ * @warning You ***MUST NOT*** call this function once the thread @p mc refers
+ *          to using the mutex @p mc refers to again (not counting the call
+ *          to @ref mutex_lock_cancelable @p mc was used in).
```
[the remark this answers is not in the document]
Comment (maribu) [reply core/include/mutex.h:276]: Me neither! :-D But I remember was I was *trying* to warn about.

**Comment by kaspar030** [comment]
> My use case was waiting for new data on UART by blocking on a mutex, but being able to inject commands from outside and unblocking the mutex for that.

Can't you just unlock the mutex in that case?

File: core/include/mutex.h [hunk 14]
```diff
@@ -215,6 +261,68 @@ void mutex_unlock(mutex_t *mutex);
...
+ * @note    It is safe to call this function from IRQ context, e.g. from a timer
+ *          interrupt.
+ * @note    It is safe to call this function more than once on the same @p mc
+ *          while it is still valid (see the warning above). The first call will
+ *          cancel the operation and subsequent calls will have no effect.
+ *
+ * @details If @p thread is currently running (or pending), a subsequent call
+ *          from @p thread to @ref mutex_lock_cancelable will - under certain
+ *          conditions - also fail
```
[the remark this answers is not in the document]
Comment (maribu) [reply core/include/mutex.h:285]: This is an artifact from a previous implementation of this PR. It will now reliably fail. I re-sync the documentation with the implementation.

**Comment by maribu** [comment]
> > My use case was waiting for new data on UART by blocking on a mutex, but being able to inject commands from outside and unblocking the mutex for that.
> 
> Can't you just unlock the mutex in that case?

I agree that this sounds a bit hacky. As both @kaspar030 and @fjmolinas agree on not making these call first class public API citizens, I will add a note that they aren't.

But I would still like to expose them for other RIOT internal users and brave enough out of tree users. In addition to canceling the signal from ISR, that @benpicco already mentioned, this could also be useful for aborting an action on a shared resource.

**Comment by fjmolinas** [comment]
> > > My use case was waiting for new data on UART by blocking on a mutex, but being able to inject commands from outside and unblocking the mutex for that.
> > 
> > 
> > Can't you just unlock the mutex in that case?
> 
> I agree that this sounds a bit hacky. As both @kaspar030 and @fjmolinas agree on not making these call first class public API citizens, I will add a note that they aren't.
> 
> But I would still like to expose them for other RIOT internal users and brave enough out of tree users. In addition to canceling the signal from ISR, that @benpicco already mentioned, this could also be useful for aborting an action on a shared resource.

Sounds like a nice compromise for me :)

File: core/include/mutex.h [hunk 15]
```diff
@@ -215,6 +269,71 @@ void mutex_unlock(mutex_t *mutex);
...
+ * @note    This function is considered internal. Out of tree users should be
+ *          aware that breaking API changes or removal of this API without
+ *          an deprecation period might happen.
+ *
+ * @pre     @p mc is used to cancel at most one call to
+ *          @ref mutex_lock_cancelable. (You can reinitialize the same memory
+ *          to safely reuse it.)
+ * @warning You ***MUST NOT*** call this function once the thread @p mc refers
+ *          to uses the mutex @p mc refers to again (not counting the call
+ *          to @ref mutex_lock_cancelable @p mc was used in).
```
Comment (fjmolinas) [inline core/include/mutex.h:288]: I'm sorry but this is still too confusing to me, does phrasing like this conserve the meaning?

```suggestion
 * @warning You ***MUST NOT*** call this function once the thread referred by @p mc
 *          re-uses the mutex object referred by @p mc (not counting the call
 *          to @ref mutex_lock_cancelable @p mc was used in).
```

Comment (maribu) [reply core/include/mutex.h]: Applied and squashed in

**Review by fjmolinas — APPROVED** [review]
Second ACK, although the new functions are exposed they are marks as internal and documentation properly warns their usage.

**Comment by aabadie** [comment]
2 ACKS, go!

**Comment by maribu** [comment]
Hooray! Thanks everyone!
